lib/services/suppressions-service.js

```diff
@@ -0,0 +1,282 @@
+                if (!unused[relativeFilePath][ruleId]) {
+                    continue;
+                }
+
+                const suppressionsCount = unused[relativeFilePath][ruleId].count;
+                const violationsCount = violationsByRule[ruleId].count;
+
+                // Suppress messages if the number of violations is less or equal to the suppressions count
+                if (violationsCount <= suppressionsCount) {
+                    SuppressionsService.suppressMessagesByRule(result, ruleId);
+                }
+
+                // Update the count to match the new number of violations, otherwise remove the rule entirely
+                if (violationsCount < suppressionsCount) {
+                    unused[relativeFilePath][ruleId].count = suppressionsCount - violationsCount;
+                } else {
+                    delete unused[relativeFilePath][ruleId];
+                }
+            }
+
+            // Cleanup files with no rules
+            if (Object.keys(unused[relativeFilePath]).length === 0) {
+                delete unused[relativeFilePath];
+            }
+        }
+
+        // Remove entries from unused that were not included in the results
+        for (const file in unused) {
+            if (!filtered.some(result => this.getRelativeFilePath(result.filePath) === file)) {
+                delete unused[file];
+            }
+        }
+
+        // Recalculate stats for the updated results
+        for (const result of filtered) {
+            Object.assign(result, calculateStatsPerFile(result.messages));
+        }
+
+        return {
+            results: filtered,
+            unused
+        };
+    }
+
+    /**
+     * Loads the suppressions file.
+     * @throws {Error} If the suppressions file cannot be parsed.
+     * @returns {Promise<SuppressedViolations>} The suppressions.
+     */
+    async load() {
+        try {
+            const data = await fs.promises.readFile(this.filePath, "utf8");
+
+            return JSON.parse(data);
+        } catch (err) {
+            if (err.code === "ENOENT") {
+                return {};
+            }
+            throw new Error(`Failed to parse suppressions file at ${this.filePath}`);
+        }
+    }
+
+    /**
+     * Updates the suppressions file.
+     * @param {SuppressedViolations} suppressions The suppressions to save.
+     * @returns {Promise<void>}
+     * @private
+     */
+    save(suppressions) {
+        return fs.promises.writeFile(
+            this.filePath,
+            JSON.stringify(suppressions, null, 2)
+        );
+    }
+
+    /**
+     * Counts the violations by rule, ignoring warnings.
+     * @param {LintMessage[]} messages The messages to count.
+     * @returns {Record<string, number>} The number of violations by rule.
+     */
+    static countViolationsByRule(messages) {
+        return messages.reduce((totals, message) => {
+            if (message.severity === 2 && message.ruleId) {
```

Comment:
I think a test that messages not related to rules are not included in suppressions, mentioned in https://github.com/eslint/eslint/pull/19159#discussion_r1982913200, is still missing.

Reply:
@mdjermanovic The test file was updated already to include a warning, and all related tests were updated to assert the warning message is being displayed.